tests/unit/common/eeprom/eeprom_access.cpp

```diff
@@ -0,0 +1,96 @@
+        throw std::out_of_range("data too large");
+    }
+    // add 4 to size,because we will add crc at the end of the data
+    std::array<uint8_t, 3> header { MSGPACK_EXT8, static_cast<unsigned char>(data.size() + 4), CUSTOM_ITEM_ID };
+
+    uint32_t crc = crc32_calc(header.data(), header.size());
+    crc = crc32_calc_ex(crc, data.data(), data.size());
+
+    uint16_t size_of_item = header.size() + data.size() + sizeof(crc);
+
+    if ((size_of_item + first_free_space) >= write_stop) {
+        Cleanup();
+    }
+
+    // persist data in eeprom
+    // write stop first, then crc,then data and last write the header, because we want to rewrite stop as the last thing in eeprom
+    // we want to write the stop only if we are writing to new
+    st25dv64k_user_write(size_of_item + first_free_space, LAST_ITEM_STOP);
+    st25dv64k_user_write_bytes(first_free_space + size_of_item - sizeof(crc), &crc, sizeof(crc));
+    st25dv64k_user_write_bytes(first_free_space + header.size(), data.data(), data.size());
+    st25dv64k_user_write_bytes(first_free_space, header.data(), header.size());
+    first_free_space += size_of_item;
+}
+std::optional<std::vector<uint8_t>> EepromAccess::load_item(uint16_t address) {
+    std::array<uint8_t, 3> header { 0, 0, 0 };
+    st25dv64k_user_read_bytes(address, header.data(), header.size());
+
+    if (header[0] != MSGPACK_EXT8 || header[2] != CUSTOM_ITEM_ID) {
+        return std::nullopt;
+    }
+    std::vector<uint8_t> data;
+    // add 4 to load also the crc
+    data.resize(header[1]);
+    st25dv64k_user_read_bytes(address + header.size(), data.data(), data.size());
+
+    uint32_t crc = crc32_calc(header.data(), header.size());
+    crc = crc32_calc_ex(crc, data.data(), data.size() - sizeof(crc));
+
+    uint32_t crc_read = *(uint32_t *)((data.data() + data.size() - sizeof(crc)));
+
+    if (crc != crc_read) {
+        return std::nullopt;
+    }
+    // remove crc from the data
+    data.resize(data.size() - sizeof(crc));
+    return data;
+}
+
+void EepromAccess::Cleanup() {
+    // TODD make this method
+}
+void EepromAccess::Init() {
+    index.clear();
+    struct IndexComparator {
+        bool operator()(const std::pair<uint32_t, uint16_t> &a, const std::pair<uint32_t, uint16_t> &b) {
+            return a.first == b.first;
+        }
+    };
+    uint16_t address = START_OFFSET;
+    for (auto item = load_item(address); item.has_value(); item = load_item(address)) {
+        auto data = item.value();
+        uint32_t key = msgpack::unpack<Key>(data).key;
+        // push back and sort
+        index.emplace_back(key, address);
+        std::push_heap(index.begin(), index.end());
+        address += item.value().size() + HEADER_SIZE + CRC_SIZE;
+    }
+    std::sort_heap(index.begin(), index.end());
+    // reversing it because if we used backward iterator we would have the unwanted indexes in front of the ones we want
+    std::reverse(index.begin(), index.end());
```

Comment:
Can you explain what this whole thing does and what's the index even for? Didn't we kind of want to mirror the values inside the RAM instead? Why sorting? Isn't it overcomplicating things?